### ISSUE TITLE
Add the `orderBy` parameter to list methods`

### ISSUE THREAD

**Issue by andrii-bodnar** [crowdin/crowdin-api-client-java#235]
The Crowdin API has been enhanced with a new `orderBy` parameter in a number of list methods. It allows sorting the results by a specific field in ascending or descending order.

The list of methods that support the new parameter:

- List projects
- List branches
- List directories
- List files
- List strings
- List Translation Approvals
- List Language Translations
- List String Translations
- List String Comments
- List Screenshots
- List Concepts
- List Glossaries
- List Terms
- List TMs
- List TM Segments
- List Tasks
- List User Tasks
- List Project Members
- List Labels
- List Groups (Enterprise only)
- List Users (Enterprise only)
- List Teams (Enterprise only)

**References:**
- [API reference](https://developer.crowdin.com/api/v2/#section/Introduction)
- [Enterprise API reference](https://developer.crowdin.com/enterprise/api/v2/)
- [Sorting rules](https://developer.crowdin.com/api/v2/#section/Introduction/SortingAI)

**Comment by halfwind22** [comment]
@andrii-bodnar Starting out with open source contributions, I would like to take this up - by when would you want this to be done?

**Comment by andrii-bodnar** [comment]
@halfwind22 the new parameter should be added to the methods listed in the description.

For example: the following is the corresponding function for the "List project" method - https://github.com/crowdin/crowdin-api-client-java/blob/master/src/main/java/com/crowdin/client/projectsgroups/ProjectsGroupsApi.java#L122

**Comment by halfwind22** [comment]
Yes @andrii-bodnar , I absolutely get your point - my question was more around how would the request url would look like `https://<hostname>/api/<version>/resource?orderBy=createdAt desc,name,priority asc,title desc` probably?

**Comment by andrii-bodnar** [comment]
@halfwind22 yes

**Comment by halfwind22** [comment]
Thanks, please assign this to me.

**Comment by halfwind22** [comment]
@andrii-bodnar I propose to allow users to pass in a Linked Hash map for specifying ordering key and the corresponding ordering sequence(ASC/DESC - enums)  . Or would it be better to allow the ordering specification in plaintext. My concern for the second approach is that an incorrect string might be passed , like `?orderBy name,id dsc ` or maybe missing out on a comma.

**Comment by andrii-bodnar** [comment]
@halfwind22 could you please provide some examples of what this might look like? I agree that passing raw parameters is unsafe and error-prone.

**Comment by halfwind22** [comment]
```
// Client side code
Map<,SortOrder> orderByMap = new LinkedHashmap<String,SortOrder>();
orderByMap.of(name, SortOrder.DESC)
orderByMap.of(id, SortOrder.ASC)

// Method Signature 
listProjects(Long groupId, Integer hasManagerAccess, Integer limit, Integer offset, LinkedHashMap<String, SortOrder> orderByMap)
```

@andrii-bodnar Above is how regular users would invoke the method. Now users might also want to just skip the SortOrder (which makes it ASC order) , so we might need to have a wrapper around the **put()** of the **Map** interface, to handle that scenario. Same goes for duplicate keys. Thoughts??

**Comment by andrii-bodnar** [comment]
@halfwind22 I'm also curious about how to make the sort order value optional.

I don't think we should care about duplicate values, it's too complicated for this feature.

**Comment by halfwind22** [comment]
A wrapper over the **put()** method of Map interface should suffice. A simpler solution would be user entering a key with no value (null type) into the map and then we interpret that null as ASCENDING. We could also probably ignore it ,because the backend REST endpoint can still work with query params coming without a sorting order  like `?orderBy name,id desc ` .

**Comment by andrii-bodnar** [comment]
Sounds good to me 🚀

**Comment by halfwind22** [comment]
Hello @andrii-bodnar , a couple of endpoints are available to only enterprise customers, while I am on a free plan. Can I get access to some test account (or temporarily elevating my account privileges) , for testing those functionalities?

**Comment by andrii-bodnar** [comment]
Hello @halfwind22, you can create an Enterprise organization here - https://accounts.crowdin.com/workspace/create

It will provide you with a 30-day trial period.

**Comment by andrii-bodnar** [comment]
@halfwind22 Let's proceed without checking for each specific endpoint. The server will return an error if a wrong key is passed.

**Comment by halfwind22** [comment]
@andrii-bodnar Done with the dev part of the code change,  for unit tests, we need to change the JSON content, by adding more than response to account for list of items in sorted order. Can I go ahead with those? Also do you have a better approach for unit testing this feature?

**Comment by andrii-bodnar** [comment]
@halfwind22, there is definitely no need to change all the tests for the list methods. It would be enough to add a separate test for each resource with the order parameter passed.

**Comment by halfwind22** [comment]
Didnt quite get your point- so let me elaborate: In line 106 of the class [https://github.com/crowdin/crowdin-api-client-java/blob/master/src/test/java/com/crowdin/client/projectsgroups/ProjectsGroupsApiTest.java](https://github.com/crowdin/crowdin-api-client-java/blob/master/src/test/java/com/crowdin/client/projectsgroups/ProjectsGroupsApiTest.java), we have the test for **listProjects** . Currently it is designed to accept only 4 arguments, so it gives a compilation error in the absence of the 5th param(orderBy) that is present in the actual method.
Also , I see the that mock responses, make use of just one project, which isn't enough to simulate the orderBy behavior, because even if we were to pass null, the default ordering must happen based on 'id'. 
Are you saying that I leave the current test untouched and instead create a new test with multiple projects, so as to simulate ordering behavior in case of a list of projects?

**Comment by andrii-bodnar** [comment]
@halfwind22 the listProjects should not fail if the `orderBy` is not present in the request since it is an optional parameter. Also, there is no need to simulate the resource order in the actual response. It would be sufficient to check that the method accepts the new parameter correctly and adds it to the request query.

**Comment by halfwind22** [comment]
@andrii-bodnar I believe that users might have different ordering preferences for each endpoint, and hence it would be a better option to have different objects for each ordering set. Do you think that users should be having only a single ordering preference at a time, and then they clear and recreate them when a new strategy is required( this enables us to make the map of order(key,strategy) as static)?

**Comment by andrii-bodnar** [comment]
Hi @halfwind22, do you have any updates on this?

**Comment by halfwind22** [comment]
Was caught up with some pressing issues at work, will send PR soon.

**Comment by halfwind22** [comment]
A bit weird ask: I am using the Eclipse IDE and it comes with a built in code style formatter. So after making changes, I accidentally formatted the file, and that means some of the formatting that was there originally is now changed. By formatting I mean tabs, new line ,etc. Could you please send me the checkstyle.xml of whatever formatter you are using? Otherwise the PR might highlight a lot of non-code changes too, leaving you confused.

**Comment by andrii-bodnar** [comment]
@halfwind22 I don't have a `checkstyle.xml` file with the formatting rules. Please disable automatic formatting for the project and don't include these changes in the PR.

**Comment by Sprokof** [comment]
Hi, does this issues is ongoing? If yes, i'we like to implement it

**Comment by halfwind22** [comment]
Hi @Sprokof , I was done with the dev part long ago. However while writing the unit tests, I started facing some weird JUnit and Mockito issues with my IDE, and haven't got much time to look into this thereafter.Would you be interested in working on my fork? Need to add unit test cases.

**Comment by Sprokof** [comment]
@halfwind22, yes, I can work on that

**Comment by halfwind22** [comment]
@Sprokof Cool.  Please confirm if you can access this branch, you should be able to see the changes.

https://github.com/halfwind22/crowdin-api-client-java/tree/feature/add-orderby-parameter

**Comment by korzol** [comment]
After all is this issue still open? Can I get it?

**Comment by korzol** [comment]
@andrii-bodnar Thanks for additional info. I will check how it was implemented in dotnet client. Kindly assign this issue to me

**Comment by andrii-bodnar** [comment]
@Korzol Thank you! Some work had already been done, please take a look at the previous discussion in the comments. Also, the methods listed in the task description might be irrelevant (the description may not include all the methods that accept `orderBy`)

**Comment by korzol** [comment]
@andrii-bodnar Sure! I’ll first check the .NET implementation, then review the work that’s already been done. After that, if everything looks good, I’ll start coding. Also, I’ll go through the documentation of all API list methods to ensure no methods are missed.
